File: Problem03/Problem03.go
package main

import (
	"fmt"
)

func main() {
	s := "abceeabdcxbb"
	fmt.Println(lengthOfLongestSubstring(s))
}

func lengthOfLongestSubstring(s string) int {
	myMap := map[rune]int{}
	maxLength := 0
	left := 0

	for right, char := range s {
		if index, found := myMap[char]; found && index >= left {
			fmt.Println("index: ", index, "left: ", left)
			left = index + 1
		}
		myMap[char] = right
		currentLength := right - left + 1
		if currentLength > maxLength {
			maxLength = currentLength
		}
	}

	return maxLength
}

File: Problem03/problem03.ts
function lengthOfLongestSubstring(s: string): number {
    const myMap: Map<string, number> = new Map();
    let maxLen = 0;
    let left = 0;

    for (let right = 0; right < s.length; right++) {
        const char = s[right];
        if (myMap.has(char) && myMap.get(char)! >= left) {
            left = myMap.get(char)! + 1;
        }
        myMap.set(char, right);
        const len = right - left + 1;
        if (len > maxLen) {
            maxLen = len;
        }
    }

    return maxLen;
};

const s = "abceeabdcxbb";
console.log(lengthOfLongestSubstring(s)); // 3